# SQL запросы

**Контекст задачи:**\
Онлайн-маркетплейс Заберу позволяет пользователям со всей России приобрести товары ручной работы от небольших производителей. Маркетплейс проводит рекламные акции длительностью 1 день. Маркетплейс хочет узнать насколько эффективны рекламные кампании.

**Источник данных:**\
По ссылке   https://yadi.sk/d/pAWR5UW4qhpJSg файлы с описанием таблиц и параметрами для подключения к базе данных. Создайте архив с итоговыми запросами и результатами и отправьте его в тренажере.

**Задание:**
- Посчитайте затраты на каждый рекламный источник.
- Посчитайте суммарную выручку, максимальную и минимальную выручку от заказов, сделанных с настольных ПК (тип устройства desktop) пользователями, которые хотя бы раз перешли из рекламного источника, на который было потрачено больше всего денег. Сделайте выводы.
- Введём новую метрику эффективности рекламы — дневные затраты на рекламу на пользователя по источнику. Она рассчитывается для каждого отдельного дня и равна объёму денежных средств, которые были израсходованы на то, чтобы пользователь перешёл по рекламе на сайт из данного рекламного источника. Иными словами, метрика для каждого дня равна отношению рекламных затрат по данному источнику к количеству посетителей, перешедших на сайт с данного источника.\
Рассчитайте для каждого дня введённую метрику.

Считайте, что целевое действие было произведено тогда, когда сессия началась. Например, если посетитель провёл на сайте несколько минут с 23:50 14 апреля по 0:20 15 апреля, то его нужно включать в расчёт за 14 апреля.

**Описание данных:**\
Таблица visits (лог сервера с информацией о посещениях сайта):
- Uid — уникальный идентификатор пользователя
- Device — категория устройства пользователя
- Start Ts — дата и время начала сессии
- End Ts — дата и время окончания сессии
- Source Id — идентификатор рекламного источника, из которого пришел пользователь

Таблица orders (информация о заказах):
- Uid — уникальный пользователя, который сделал заказ
- Buy Ts — дата и время заказа
- Revenue — выручка Яндекс.Афиши с этого заказа

Таблица advertisment_costs (информация о затратах на маркетинг):
- source_id — идентификатор рекламного источника
- dt — дата
- costs — затраты на этот рекламный источник в этот день

In [1]:
# установка библиотеки для доступа к PostgreSQL
!pip install psycopg2

In [2]:
# Импорт библиотек
import sys
import getopt
from datetime import datetime
import pandas as pd
from sqlalchemy import create_engine

In [3]:
# настройка окружения
pd.options.display.max_columns=None
pd.options.display.float_format='{:.2f}'.format

## Подключение к БД

In [4]:
# Параметры подключения к БД

db_config = {'user': 'analyst',         # имя пользователя
             'pwd': 'I80qcMa',          # пароль
             'host': 'logdb.cdxwh80nuzgz.us-west-2.rds.amazonaws.com', # адрес сервера
             'port': 5432,              # порт подключения
             'db': 'shopinfo'}          # название базы данных

# Строка соединения с БД
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                        db_config['pwd'],
                                                        db_config['host'],
                                                        db_config['port'],
                                                        db_config['db'])

In [5]:
# Подключаемся к БД.
engine = create_engine(connection_string)

# Выполняем запрос и сохраняем результат
# выполнения в DataFrame

# Формируем sql-запросы для первой оценки данных
# загружаем только первые 50 строк
query_vizit = '''SELECT *
        FROM visits 
        LIMIT 50
        '''

query_orders = '''SELECT *
        FROM orders 
        LIMIT 50'''

query_advertisment_costs = '''SELECT *
        FROM advertisment_costs 
        LIMIT 50'''
    
df_visits = pd.io.sql.read_sql(query_vizit, con = engine, index_col = 'id')
df_orders = pd.io.sql.read_sql(query_orders, con = engine, index_col = 'id')
df_advert = pd.io.sql.read_sql(query_advertisment_costs, con = engine, index_col = 'id')

#### Визульная оценка данных в таблицах

In [6]:
df_visits.head(3)

uid   device  sourceid             startts  \
id                                                                  
0  16879256277536000000.00    touch         4 2017-12-20 17:20:00   
1    104060357244891744.00  desktop         2 2018-02-19 16:53:00   
2   7459035603376831488.00    touch         5 2017-07-01 01:54:00   

                 endts  
id                      
0  2017-12-20 17:38:00  
1  2018-02-19 17:21:00  
2  2017-07-01 01:54:00

In [7]:
df_orders.head(3)

uid               buyts revenue
id                                                       
14336 10329302124590700544.00 2017-06-01 00:10:00  $17.00
14337 11627257723692900352.00 2017-06-01 00:25:00   $0.55
14338 17903680561304199168.00 2017-06-01 00:27:00   $0.37

In [8]:
df_advert.head(3)

source_id          dt   costs
id                               
1           1  2017-06-01  $75.20
2           1  2017-06-02  $62.25
3           1  2017-06-03  $36.53

In [9]:
# Получение информации о типе данных в таблицах БД
# 
query_data_type = '''
    SELECT table_name, column_name, data_type 
    FROM INFORMATION_SCHEMA.COLUMNS 
    WHERE table_name = 'visits' OR table_name = 'orders' OR table_name = 'advertisment_costs';
    '''
# выполнение запроса к БД
df_query_1 = pd.io.sql.read_sql(query_data_type, con = engine)
# вывод результата
df_query_1.sort_values(by='table_name')

table_name column_name                    data_type
4   advertisment_costs          id                      integer
5   advertisment_costs   source_id                      integer
6   advertisment_costs          dt                         text
7   advertisment_costs       costs                        money
0               orders          id                      integer
1               orders         uid                      numeric
2               orders       buyts  timestamp without time zone
3               orders     revenue                        money
8               visits          id                      integer
9               visits         uid                      numeric
10              visits      device                         text
11              visits    sourceid                      integer
12              visits     startts  timestamp without time zone
13              visits       endts  timestamp without time zone

**Вывод:** в таблице  'advertisment_costs' поле 'dt' имеет тип 'text' вместо 'timestamp', нужно учитывать при составлении запроса.

## Решение основного задания

#### ЗАДАНИЕ 1
Посчитайте затраты на каждый рекламный источник.

In [24]:
# Задание 1
# SQL запрос - ВАРИАНТ 1 (сортировка средствами SQL)

query_1 = '''
          SELECT source_id, SUM(costs) AS sum_costs
          FROM advertisment_costs
          GROUP BY source_id
          ORDER BY sum_costs DESC
       '''
# выполнение запроса к БД
df_query_1 = pd.io.sql.read_sql(query_1, con = engine)
# вывод результата
df_query_1

source_id    sum_costs
0          3  $141,321.63
1          4   $61,073.60
2          5   $51,757.10
3          2   $42,806.04
4          1   $20,833.27
5         10    $5,822.49
6          9    $5,517.49

**Вывод:** Наибольшие затраты были на источник 3, наименьшие на источник 9.

#### ЗАДАНИЕ 2
Посчитайте суммарную выручку, максимальную и минимальную выручку от заказов, сделанных с настольных ПК (тип устройства desktop) пользователями, которые хотя бы раз перешли из рекламного источника, на который было потрачено больше всего денег. Сделайте выводы.

Я сейчас немного поправил описание задачи. Кажется, что фраза "Рассчитайте для каждого дня введённую метрику" относится только к 3-ьему заданию. Небольшой косяк форматирования тебя смутил.
Таким образом во втором задании нужно сделать подзапрос с выбором тех пользователей, что были в источнике с максимальными затратами. Все хитрости с датами там не нужны и итоговый результат не должен быть разбит по дням.

In [59]:
query_1 = '''
   SELECT 
       SUM(revenue),
       MAX(revenue),
       MIN(revenue)
   FROM
       orders
   WHERE
       uid IN (SELECT uid
               FROM visits 
                    INNER JOIN
                         (SELECT source_id,  SUM(costs) AS sum_costs
                          FROM advertisment_costs
                          GROUP BY source_id
                          ORDER BY sum_costs DESC
                          LIMIT 1) AS source
                    ON source.source_id = visits.sourceid 
               WHERE device='desktop')
       AND CAST(revenue AS NUMERIC(12,2)) > 0

'''
# выполнение запроса к БД
df_query_1 = pd.io.sql.read_sql(query_1, con = engine)
# вывод результата
df_query_1

sum      max    min
0  $13,330.09  $229.78  $0.01

**Вывод:**\
Самым дорогим источником рекламы явлется 3-й. Затраты составили 141321у.е. При этом общая выручка по переходам от него 13330 у.е. **Источник неэффективен.**

#### ЗАДАНИЕ 3
Введём новую метрику эффективности рекламы — дневные затраты на рекламу на пользователя по источнику. Она рассчитывается для каждого отдельного дня и равна объёму денежных средств, которые были израсходованы на то, чтобы пользователь перешёл по рекламе на сайт из данного рекламного источника. Иными словами, метрика для каждого дня равна отношению рекламных затрат по данному источнику к количеству посетителей, перешедших на сайт с данного источника.
Рассчитайте для каждого дня введённую метрику.

In [15]:
# Задание 3
# SQL запрос


query_3 = '''
SELECT 
    SQ_visits.startts,
    SQ_visits.sourceid,
    CAST(SQ_advert.costs / SQ_visits.count AS NUMERIC(12,2))  AS unit_cost
        
FROM
    (SELECT DATE_TRUNC('day', startts) AS startts, sourceid, COUNT(DISTINCT uid) AS count
    FROM visits 
    GROUP BY DATE_TRUNC('day', startts), sourceid) AS SQ_visits
    
    LEFT JOIN
    
    (SELECT DATE_TRUNC('day', CAST(dt AS DATE)) AS dt, source_id, SUM(costs) AS costs
    FROM advertisment_costs
    GROUP BY DATE_TRUNC('day', CAST(dt AS DATE)), source_id) AS SQ_advert
    
    ON SQ_visits.startts = SQ_advert.dt AND SQ_visits.sourceid = SQ_advert.source_id
    '''

# выполнение запроса к БД
df_query_3 = pd.io.sql.read_sql(query_3, con = engine)
# вывод результата
df_query_3.sort_values(by='startts').head(10)

startts  sourceid  unit_cost
0 2017-06-01         1      10.74
1 2017-06-01         2       7.36
2 2017-06-01         3      24.95
3 2017-06-01         4       7.67
4 2017-06-01         5       8.13
5 2017-06-01         9       3.26
6 2017-06-01        10       7.54
7 2017-06-02         1      20.75
8 2017-06-02         2      26.47
9 2017-06-02         3      14.96